device/transport/amqp/test/_amqp_test.js

```diff
@@ -4,246 +4,1345 @@
-    /*Tests_SRS_NODE_DEVICE_AMQP_06_011: [The `updateSharedAccessSignature` method shall call the `done` callback with a null error object and a SharedAccessSignatureUpdated object as a result, indicating the client does NOT need to reestablish the transport connection.]*/
-    /*Tests_SRS_NODE_DEVICE_AMQP_06_010: [The `updateSharedAccessSignature` method shall call the amqp transport `putToken` method with the first parameter audience, created from the sr of the sas signature, the next parameter of the actual sas, and a callback.]*/
-    it('saves sharedAccessSignature - invokes puttoken and passes back result indicating NOT needing to reconnect', function () {
-      var transport = new Amqp(configWithSAS);
-      sinon.stub(transport._amqp,'putToken').callsArgWith(2, null);
-      transport.updateSharedAccessSignature(simpleSas, function(err, result) {
-        assert.equal(transport._config.sharedAccessSignature, simpleSas);
-        assert.isNotOk(err);
-        assert.isFalse(result.needToReconnect);
+        // now blocked in "connecting" state
+        transport.connect(function (err, result) {
+          assert.instanceOf(result, results.Connected);
+          assert(fakeBaseClient.connect.calledOnce, 'the second stub for connect has not been called');
+          testCallback();
+        });
+
+        connectCallback(connectErr);
+      });
+
+      it('defers the call if already connected and authenticating', function (testCallback) {
+        var connectErr = new Error('cannot connect');
+        var authCallback;
+        fakeBaseClient.initializeCBS = sinon.stub().callsFake(function (done) {
+          authCallback = done;
+        });
+
+        transport.connect(function (err) {
+          fakeBaseClient.initializeCBS = sinon.stub().callsArgWith(0, null);
+        });
+
+        // now blocked in "authenticating" state
+        transport.connect(function (err, result) {
+          assert.instanceOf(result, results.Connected);
+          assert(fakeBaseClient.connect.calledOnce, 'the second stub for connect has not been called');
+          testCallback();
+        });
+
+        authCallback();
+      });
+
+      it('defers the call if disconnecting', function (testCallback) {
+        var disconnectCallback;
+        fakeBaseClient.disconnect = sinon.stub().callsFake(function (done) {
+          disconnectCallback = done;
+        });
+
+        transport.connect(function () {
+          assert(fakeBaseClient.connect.calledOnce);
+          transport.disconnect(function (done) {
+            disconnectCallback = done;
+          });
+
+          // now blocked in "connecting" state
+          transport.connect(function (err, result) {
+            assert.instanceOf(result, results.Connected);
+            assert(fakeBaseClient.connect.calledTwice);
+            testCallback();
+          });
+
+          disconnectCallback();
+        });
+      });
```

Comment:
General comment about this whole file, which I'm not going to review in detail:  All this yellow is awesome!  These kinds of tests make me so happy.  I am worried that we have all these tests which are verifying functionality that isn't captured in our requirements doc.  Yes, this has some internal knowledge (caller doesn't know what 'defer' means in this context and shouldn't care), but we could write reqs like
If connect is called while the transport is being disconnect, connect shall wait until the disconnection is complete and then re-connect.
or something like that.  

Not something we should do now, but I'm wondering if this practice would be useful in the future.

Reply:
I have been wondering how to tackle this. you're right, we need to think about it more. 

---
In reply to: [141139965](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141139965) [](ancestors = 141139965)